util/topgen/intermodule.py

```diff
@@ -69,6 +69,93 @@ def get_suffixes(ims: OrderedDict) -> (str, str):
+    # check if rsp_key
+    if rsp_key in connect:
+        # check if rsp has data
+        if req_key in connect[rsp_key]:
+            return
+        rsp_key.append(req_key)
+        return
+
+    # Add new key and connect
+    connect[req_key] = [rsp_key]
+
+
+def autoconnect(topcfg: OrderedDict):
+    """Matching the connection based on the naming rule
+    between {memory, module} <-> Xbar.
+    """
+
+    instances = topcfg["module"] + topcfg["memory"]
+
+    for xbar in topcfg["xbar"]:
+        ports = [x for x in xbar["nodes"] if x["type"] in ["host", "device"]]
+        for port in ports:
+            if port["xbar"]:
+                if port["type"] == "host":
+                    # Skip as device will add connection
+                    continue
+                add_intermodule_connection(obj=topcfg,
+                                           req_m=xbar["name"],
+                                           req_s="tl_%s" % port["name"],
+                                           rsp_m=port["name"],
+                                           rsp_s="tl_%s" % xbar["name"])
+
+            # check if name exists in ["module", "memory"]
+            ips = [x for x in instances if x["name"] == port["name"]]
+            if len(ips) == 1:
+                ip = ips[0]
+                # get the port name
+                if port["type"] == "device":
+                    ims = [
+                        x for x in ip["inter_signal_list"]
+                        if (x["package"] == "tlul_pkg" if "package" in
+                            x else False) and x["struct"] == "tl" and
+                        x["act"] == "rsp"
+                    ]
+                    assert len(ims) == 1
+                    sig_name = ims[0]["name"]
+                else:
+                    # host type
+                    ims = [
+                        x for x in ip["inter_signal_list"]
+                        if (x["package"] == "tlul_pkg" if "package" in
+                            x else False) and x["struct"] == "tl" and
+                        x["act"] == "req"
+                    ]
+                    assert len(ims) == 1
+                    sig_name = ims[0]["name"]
+
+                add_intermodule_connection(obj=topcfg,
+                                           req_m=port["name"],
+                                           req_s=sig_name,
+                                           rsp_m=xbar["name"],
+                                           rsp_s="tl_%s" % port["name"])
+                continue  # len() == 1
+
+            # if not in module, memory, should be existed in top or ext field
+            module_key = "{}.tl_{}".format(xbar["name"], port["name"])
+            if module_key not in topcfg["inter_module"]["top"] + topcfg[
+                    "inter_module"]["external"]:
+                log.error("Inter-module key {} is not in module, memory list. "
+                          "Shall be in top or ext".format(module_key))
```

Comment:
nit: maybe rephrase as
`The inter-module key {} can not be found in the module / memory lists, nor in the top / ext lists.`

Reply:
why don't we just say `can not be found in module, memory, top or ext lists`?

Reply:
Looks good. Let me revise